sqlalchemy_bigquery/base.py

```diff
@@ -246,6 +250,56 @@ def visit_insert(self, insert_stmt, asfrom=False, **kw):
             insert_stmt, asfrom=False, **kw
         )
 
+    def visit_table_valued_alias(self, element, **kw):
+        # When using table-valued functions, like UNNEST, BigQuery requires a
+        # FROM for any table referenced in the function, including expressions
+        # in function arguments.
+        #
+        # For example, given SQLAlchemy code:
+        #
+        #   print(
+        #      select([func.unnest(foo.c.objects).alias('foo_objects').column])
+        #      .compile(engine))
+        #
+        # Left to it's own devices, SQLAlchemy would outout:
+        #
+        #   SELECT `foo_objects`
+        #   FROM unnest(`foo`.`objects`) AS `foo_objects`
+        #
+        # But BigQuery diesn't understand the `foo` reference unless
+        # we add as reference to `foo` in the FROM:
+        #
+        #   SELECT foo_objects
+        #   FROM `foo`, UNNEST(`foo`.`objects`) as foo_objects
+        #
+        # This is tricky because:
+        # 1. We have to find the table references.
+        # 2. We can't know practically if there's already a FROM for a table.
+        #
+        # We leverage visit_column to find a table reference.  Whenever we find
+        # one, we create an alias for it, so as not to conlfict with an existing
+        # reference if one is present.
+        #
+        # This requires communicating between this function and visit_column.
+        # We do this by sticking a dictionary in the keyword arguments.
```

Comment:
Is it a documented feature of SQLAlchemy that it'll always use `**kwargs` to pass through between these functions? If so, can we link to such a document?

If not, I'm a bit wary of this method of communication, though I don't have any better ideas.

Reply:
That's a good question.  I've seen it so much in the code that I've just taken it for granted.  

I asked here: https://github.com/sqlalchemy/sqlalchemy/discussions/6931.

Often options are explicit keyword arguments that become `**kwargs`, usually by necessity because arguments get passed through many intermediate layers that don't care about them.

In the case, an alternative would be to add a private/`__` attribute to the alias element, which is passed by SQLAlchemy to `visit_column` as the   `'enclosing_alias'` key in `kwargs`.  :)  However, relying on our own key seemed safer to me.

I'm curious to see what response I get in the discussion.

Reply:
> So that said if you are using your own values in the **kw passed through compiler.process() and all the visit_ methods, that's fine, might be a good idea to qualify the names you use so they don't conflict with future names.

Sounds like you're doing the right thing here.